File: sdk/ruby/lib/anyvali/validation_context.rb
# frozen_string_literal: true

module AnyVali
  class ValidationContext
    attr_reader :definitions
    attr_accessor :inherited_unknown_keys, :sensitive_mode, :sensitive_transform, :sensitive_cache

    def initialize(definitions: {})
      @definitions = definitions
      @inherited_unknown_keys = nil
      @sensitive_mode = nil
      @sensitive_transform = nil
      @sensitive_cache = nil
    end
  end
end

File: sdk/csharp/src/AnyVali/ValidationContext.cs
namespace AnyVali;

/// <summary>
/// Internal context for tracking path and issues during validation.
/// </summary>
public sealed class ValidationContext
{
    public List<object> Path { get; } = new(); // string or int
    public List<ValidationIssue> Issues { get; } = new();
    public UnknownKeyMode? InheritedUnknownKeys { get; set; }
    internal string? SensitiveMode { get; set; }
    internal Func<IReadOnlyList<object>, object?, object?>? SensitiveTransform { get; set; }
    internal Dictionary<string, object?>? SensitiveCache { get; set; }

    public List<object> ClonePath() => new(Path);

    public void PushPath(object segment) => Path.Add(segment);
    public void PopPath() => Path.RemoveAt(Path.Count - 1);

    internal void InheritSensitive(ValidationContext parent)
    {
        SensitiveMode = parent.SensitiveMode;
        SensitiveTransform = parent.SensitiveTransform;
        SensitiveCache = parent.SensitiveCache;
    }
}
